tests/tests/test_preauth.py

```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/python
+# Copyright 2017 Northern.tech AS
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        https://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+from mendertesting import MenderTesting
+from common_setup import *
+from MenderAPI import adm, deviceauth, inv
+import pytest
+import json
+import logging
+
+from Crypto.PublicKey import RSA
+
+import time
+
+
+class TestPreauthBase(MenderTesting):
+    def do_test_ok_preauth_and_bootstrap(self):
+        """
+            Test the happy path from preauthorizing a device to a successful bootstrap.
+            Verify that the device/auth set appear correctly in admission API results.
+        """
+        client = get_mender_clients()[0]
+
+        # we'll use the same pub key for the preauth'd device, so get it
+        res = execute(Client.get_pub_key, hosts=client)
+        preauth_key = res[client].exportKey()
+
+        # stick an extra newline on the key - this is how a device would send it
+        preauth_key += '\n'
+
+        # preauthorize a new device
+        preauth_iddata = {"mac": "mac-preauth"}
+        # serialize manually to avoid an extra space (id data helper doesn't insert one)
+        preauth_iddata_str = "{\"mac\":\"mac-preauth\"}"
+
+        r = adm.preauth(preauth_iddata_str, preauth_key)
+        assert r.status_code == 201
+
+        # verify the device appears correctly in api results
+        devs = adm.get_devices(2)
+
+        dev_preauth = [d for d in devs if d['status'] == 'preauthorized']
+        assert len(dev_preauth) == 1
+        dev_preauth = dev_preauth[0]
+        assert dev_preauth['device_identity'] == preauth_iddata_str
+        assert dev_preauth['key'] == preauth_key
+
+        # make one of the existing devices the preauthorized device
+        # by substituting id data and restarting
+        res = execute(Client.substitute_id_data, preauth_iddata, hosts=client)
+        res = execute(Client.restart, hosts=client)
+
+        # verify api results - after some time the device should be 'accepted'
+        time.sleep(30)
```

Comment:
Increase this timeout

Reply:
what's your recommendation?

Reply:
btw this one should be instantaneous, I just restart the mender binary inside quemu (not the client itself)

Reply:
also, the tests are green. if we don't have good ideas for a precise number I'd leave it well alone for the time being.